File: 257 Binary Tree Paths.py
from copy import deepcopy
'''
Binary Tree Paths
https://leetcode.com/problems/binary-tree-paths/
'''
NEXT = "->"

def rec(root,buf,pool):
    if root is None:
        return
    if root.left is None and root.right is None:
        pool.append(deepcopy(buf) + str(root.val))
    elif root.left is not None and root.right is None:
        rec(root.left, buf + str(root.val) + NEXT, pool)
    elif root.left is None and root.right is not None:
        rec(root.right,buf + str(root.val) + NEXT, pool)
    else:
        newBuf = deepcopy(buf)
        rec(root.left, buf + str(root.val) + NEXT, pool)
        rec(root.right, newBuf + str(root.val) + NEXT, pool)

def binaryTreePaths(root):
    pool = []
    buf = ""
    rec(root,buf,pool)
    return pool

File: 257 Binary Tree Paths.ts
/*
257. Binary Tree Paths
https://leetcode.com/problems/binary-tree-paths/

Given a binary tree, return all root-to-leaf paths.

Note: A leaf is a node with no children.

Example:

Input:

   1
 /   \
2     3
 \
  5

Output: ["1->2->5", "1->3"]

Explanation: All root-to-leaf paths are: 1->2->5, 1->3
*/

function binaryTreePaths(root: TreeNode | null): string[] {
    let pool: string[] = []

    function topdown(node: TreeNode, parents: number[]): void {
        if (node.left !== null && node.right !== null) {
            topdown(node.left, parents.concat([node.val]))
            topdown(node.right, parents.concat([node.val]))
        }
        else {
            parents.push(node.val)
            if (!node.left && !node.right) {
                pool.push(parents.join('->'))
            }
            else if (node.left !== null) {
                topdown(node.left, parents)
            }
            else if (node.right !== null) {
                topdown(node.right, parents)
            }
        }
    }

    if (root !== null) {
        topdown(root, [])
    }
    return pool
}
